### ISSUE TITLE
API and Behaviour Changes for next major release

### ISSUE THREAD

**Issue by pjbgf** [go-git/go-git#910]
Potential API changes for the next major release. This changes will be merged into the development branch [main](https://github.com/go-git/go-git/tree/main), before they are released on the upcoming `v6.0.0` release.

#### Changes:
- [ ] SHA256 Support: The support for SHA256 and SHA1 in parallel may require API changes on `plumbing.Hash` or where it is currently used.
  - xref: https://github.com/go-git/go-git/issues/706 https://github.com/go-git/go-git/issues/899
- [ ] Git config management (TBC)
  - xref: https://github.com/go-git/go-git/issues/395
- [ ] `worktree_status.go`
    - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/worktree_status.go#L273
    - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/worktree_status.go#L372
- [ ] `difftree.go`
    - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/plumbing/object/difftree.go#L63C1-L63C1
- [ ] `options.go`
    - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/options.go#L711
- [ ] `repository.go`
    - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/repository.go#L561
    - https://github.com/go-git/go-git/blob/0a1c5ab118998a22ef88d8e6c42970197d77413c/repository.go#L426
    - https://github.com/go-git/go-git/blob/0a1c5ab118998a22ef88d8e6c42970197d77413c/repository.go#L439
- [ ] `server.go`
    - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/plumbing/transport/server/server.go#L92
- [x] `commitgraph`
  - https://github.com/go-git/go-git/pull/854#discussion_r1342188779
  - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/plumbing/format/commitgraph/memory.go#L10
  - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/plumbing/format/commitgraph/doc.go#L3
  - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/plumbing/format/commitgraph/encoder.go#L14
  - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/plumbing/format/commitgraph/file.go#L16
  - https://github.com/go-git/go-git/blob/cf196ea5ba6c8b16712a9868cc2d4cca4f287ae4/plumbing/format/commitgraph/commitgraph.go#L12

- [ ] make [osfs.BoundOS](https://github.com/go-git/go-billy/blob/master/osfs/os_bound.go) the default osfs for `Plain` operations.
- [ ] All tags not being returned by default. https://github.com/go-git/go-git/pull/1459
- [ ] New transport API
- [ ] [config.Version_X](https://github.com/ferhatelmas/go-git/blob/1afa973bd311817bba8ce515e017ec37e3287872/plumbing/format/config/format.go#L19) -> [config.VersionX](https://github.com/ferhatelmas/go-git/blob/fc79cd74a228334b0ce86e6fe456f60f0b1eed10/plumbing/format/config/format.go#L19) as part of https://github.com/go-git/go-git/pull/1771.
- [ ] Config validation no longer rejects `branch.*.merge` values without the `ref/` prefix. #1923

Potential breaking changes:

- [ ] `repository.go`
  - https://github.com/go-git/go-git/blob/0a1c5ab118998a22ef88d8e6c42970197d77413c/repository.go#L440
- [ ] Git config management (TBC)
  - xref: https://github.com/go-git/go-git/issues/395
- [ ] Need to `.Close()` repositories as part of https://github.com/go-git/go-git/pull/1906.

**Comment by davidalpert** [comment]
Would it be appropriate to consider v6.0.0 as an opportunity to replace the current pattern of duplicating `Function(..)` and `FunctionWithOptions(...)` with a functional options pattern (like [this example](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis
)) that provides better extensibility into the future?

So instead of
```
type InitOptions struct {
	// The default branch (e.g. "refs/heads/master")
	DefaultBranch plumbing.ReferenceName
}

func Init(s storage.Storer, worktree billy.Filesystem) (*Repository, error) {
	options := InitOptions{
		DefaultBranch: plumbing.Master,
	}
	return InitWithOptions(s, worktree, options)
}

func InitWithOptions(s storage.Storer, worktree billy.Filesystem, options InitOptions) (*Repository, error) {
     ...
}
```
We might support
```
type InitOptions struct {
	// The default branch (e.g. "refs/heads/master")
	DefaultBranch plumbing.ReferenceName
}

func NewInitOptions() InitOptions {
    return InitOptions{
		DefaultBranch: plumbing.Master,
	}
}

type WithInitOptions = (func *InitOptions)

func InitWithDefaultBranch(br string) WithInitOptions {
    return func(o *InitOptions) {
        o.DefaultBranch = br
    }
}

func Init(s storage.Storer, worktree billy.Filesystem, ofns ...WithInitOptions) (*Repository, error) {
    o := NewInitOptions()
    for _, fn := range ofns {
        fn(&o)
    }

     ...
}
```

This pattern supports calling Init from outside the `git` package like this
```
repo, err := git.Init(storer, worktreeFS)
```
And like this
```
repo, err := git.Init(storer, worktreeFS, git.InitWithDefaultBranch(plumbing.Main))
```

**Comment by pjbgf** [comment]
All the changes here should make their way into the [v6 migration page](https://go-git.github.io/docs/tutorials/migrating-from-v5-to-v6/).
